### PR TITLE
librewolf: init at 97.0-2

### LINKED ISSUE #94918: Package LibreWolf
**Project description**
A fork of Firefox, focused on privacy, security and freedom.

**Metadata**

* homepage URL: https://librewolf-community.gitlab.io/
* source URL: https://gitlab.com/librewolf-community/browser
* license: Mozilla Public License 2.0
* platforms: linux, darwin, windows

### PR DESCRIPTION
**Pull request by squalus**
###### Motivation for this change
This PR adds a package for the LibreWolf browser. LibreWolf is a fork of Firefox, focused on privacy, security and freedom. Fixes #94918

There are a few small changes to the existing Firefox expressions. These changes ~should not trigger any rebuilds~ will trigger a rebuild only of the Firefox wrappers (not the compiled artifacts).

An updater script is included. Updating still requires some manual analysis work. The ~patch list~, flags, and file replacements can't yet be automatically read from upstream.

Firefox updates should not be delayed by the presence of the Librewolf expressions. The versioning is disconnected. There is a small maintenance burden on Firefox: `common.nix` and `wrapper.nix` need to remain compatible with the version used by LibreWolf. Since they are also used for earlier ESR releases, I don't anticipate that to be much of a problem.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Kranzes** [comment]
Could you also create `librewolf-bin`, similar to`firefox-bin`?
Many users use `firefox-bin` instead of just `firefox` because of this issue: https://github.com/NixOS/nixpkgs/issues/76484

**Comment by squalus** [comment]
Yeah, we could make a `librewolf-bin` from the upstream AppImage in some future work.

**Comment by Kranzes** [comment]
> Yeah, we could make a `librewolf-bin` from the upstream AppImage in some future work.

Appimages don't work very well, when having only pipewire (with all emulation enabled) the audio does not work, they also tend to crash sometimes when opening file pickers. Maybe we can use a different binary that isn't an appimage?

**Comment by squalus** [comment]
I'm not sure what this error means.

```error: cannot read '/build/store/ys0zw3d95amfhzmafv15v3xp3cmlxgxi-source/distribution/policies.json', since path '/build/store/5lqaz3dri6hfnpiwr18w7ggglmgcgpxx-source.drv' is not valid```

It does reproduce on my system when running the `nix-env` command from the release checks script. It builds fine with `nix-build` though.

**Comment by squalus** [comment]
For a binary packages, we're limited to what upstream provides. That looks like AppImage, Flatpak, and distro-specific packages.

**Comment by Kranzes** [comment]
> > > Yeah, we could make a `librewolf-bin` from the upstream AppImage in some future work.
> > 
> > 
> > Appimages don't work very well, when having only pipewire (with all emulation enabled) the audio does not work, they also tend to crash sometimes when opening file pickers. Maybe we can use a different binary that isn't an appimage?
> 
> For binary packages, we're limited to what upstream provides. That looks like AppImage, Flatpak, and distro-specific packages.

**Comment by Kranzes** [comment]
Missclicked, sorry.
Take a look at this:
https://gitlab.com/api/v4/projects/12829184/packages/generic/librewolf/95.0.2-1/LibreWolf-95.0.2-1.x86_64.tar.bz2
https://gitlab.com/api/v4/projects/12829184/packages/generic/librewolf/95.0.2-1/LibreWolf-95.0.2-1.aarch64.tar.bz2

These are standalone precompiled binaries, I managed to run them fine using `steam-run`.

**Comment by squalus** [comment]
I think I understand the problem - I'm reading a downloaded file during eval time. I'm going to change the firefox wrapper to accept filenames as well as verbatim text/attrSets for extraPrefs/extraPolicies. That will probably trigger a rebuild of existing firefox wrappers, but those are the fast part of the build.

**Comment by squalus** [comment]
This is working ok for me. I think it's ready for testing and review.

**Comment by jtrees** [comment]
Is this blocked by something?

**Comment by mweinelt** [comment]
Likely nothing is happening here because the firefox maintainers are not interested in yet another fork that reuses its `common.nix` that may result in breakage when one of the packages updates it in a incompatible way.

**Comment by jtrees** [comment]
FYI: I have been daily-driving this for several weeks. Works great. I'm not equipped to evaluate the code changes though.

**Comment by onny** [comment]
Builds and is working fine for me, thank you for this really great work :)

**Comment by squalus** [comment]
Thanks for the testing.

I rebased onto master and squashed into a single commit.

**Comment by Kranzes** [comment]
I would suggest formatting all the files with nixpkgs-fmt to make them look nicer.

File: pkgs/top-level/all-packages.nix
Comment (mweinelt) [inline pkgs/top-level/all-packages.nix:25725]:
```suggestion
  };
  librewolf-wayland = wrapFirefox librewolf-unwrapped { forceWayland = true; };
```

A wayland variant would be appreciated.

Comment (squalus) [reply pkgs/top-level/all-packages.nix:25725]: Added the wayland variant

**Comment by mweinelt** [comment]
- [x] YouTube (also Live)
- [x] Webcam (Pipewire, xdg-desktop-portal-wlroots)
- [x] Microphone (Pipewire)
- [x] Screensharing (xdg-desktop-portal-wlroots, wlroots)
- [ ] WebGL (disabled)

**Review by mweinelt — APPROVED** [review]

**Pull request merged**
